### PR TITLE
Milestone 5: Friendships v1

### PR DESCRIPTION
**Pull request by ElijahTobs**
Hi TSE,

In this milestone, the requested features have been were implemented:

A logged-in user is able to:

- send friend request
- accept or reject friend request
- see pending invitations 
- see if a user is already a friend 

Kindly review.

### CODE REVIEW

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello @ElijahTobs 
### You have done well so far :clap: However, you have a few issues to fix before the project can be approved.
- [x] You need to add unit tests cases for the app
- [x] A user can invite friends
- [x] A user can see friendship requests
- [x] A user can reject friendship
- [x] There is a button to invite user to friendship on users list page and on single user page
### Add these functionalities and make the required changes. Submit for another code review. Good luck.
- Happy hacking!

By Raymond
Slack @iamraymond

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -1,9 +1,23 @@
...
+      <% if current_user.friend_invites(@user) %>
+        <button class="btn btn-warning"> Invitation sent </button>
+      <% elsif current_user.receive_invitation(@user) %>
+        <%= link_to 'Accept', accept_path(user_id: @user.id), class:"btn btn-primary" %>
+        <%= link_to 'Decline', reject_path(user_id: @user.id), method: :delete, class:"btn btn-danger" %>
+      <% elsif !current_user.friend?(@user) && current_user.id != @user.id %>
+        <%= link_to 'Add Friend', invite_path(user_id: @user.id) , class:"btn btn-success" %>
+      <% elsif current_user.friend?(@user)  %>
+        <button class="btn btn-warning"> A friend </button>
+      <% end %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: remove logic from views and move them into view helpers

File: app/views/users/index.html.erb [hunk 2]
```diff
@@ -1,12 +1,31 @@
...
+          <% if current_user.friend_invites(user) %>
+            <button class="btn btn-warning"> Invitation sent </button>
+          <% elsif current_user.receive_invitation(user)%>
+            <%= link_to 'Accept', accept_path(user_id: user.id), class:"btn btn-primary" %>
+            <%= link_to 'Decline', reject_path(user_id: user.id), method: :delete, class:"btn btn-danger" %>
+          <% elsif !current_user.friend?(user) && current_user.id != user.id %>
+            <%= link_to 'Add Friend', invite_path(user_id: user.id) , class:"btn btn-success" %>
+          <% elsif current_user.friend?(user)  %>
+            <button class="btn btn-warning"> A friend </button>
+          <% end %>
```
Comment (maelfosso) [inline app/views/users/index.html.erb]: Move logic into view helpers

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you applying the suggested changes.
However, there are some missing requirements

1. It's required that the posts timeline must include posts written by logged-in user and their friends, as indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png). It's not the case in your web application
Use this
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

2. Please, there are no unit and integration tests, as required [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png)

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: app/views/users/index.html.erb (hunk 2 above)
Comment (ElijahTobs) [reply app/views/users/index.html.erb]: Logic has been moved to helpers.

Thanks

File: app/views/users/show.html.erb (hunk 1 above)
Comment (ElijahTobs) [reply app/views/users/show.html.erb]: Logic has been moved to helpers.

Thanks

**Comment by ElijahTobs** [comment]
Hi TSE,

When I try to run `rspec`, I get that error message. I've checked many answers on stack-overflow but none worked for me.
I would appreciate if you can help me with this:

spec -error

Warm regards,
Elijah

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello @ElijahTobs 
### You have done well so far :clap: However, your tests are failing. Kindly fix this:-
- [x] Check your rspec configuration, use this [link](https://hixonrails.com/ruby-on-rails-tutorials/ruby-on-rails-testing-rspec-configuration/) or this [link](https://dzone.com/articles/rails-rspec-setup-1)
### Make the changes and submit for another code review. Good luck.
- Happy coding!

**Review by MuriloRoque — APPROVED** [review]
# Status: Approved :green_circle: 

### Hello @ElijahTobs :wave: 

### Thank you for making all of the required changes, your project is approved :clap: :clap: :clap: 

### There's only one minor detail you should fix: In your README file, you misspelled 'rspec' in `rpsec --format documentation` :wink: 

### You may merge, good luck with your next projects :rocket: 

Murilo Roque Paiva da Silva

- Github: [@muriloroque](https://github.com/MuriloRoque)
- Linkedin: [MuriloRoque](https://www.linkedin.com/in/murilo-roque-b1268741/)

[Twitter Follow](https://twitter.com/MuriloRoquePai1)

**Pull request merged**